oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -0,0 +1,348 @@
+  enforce a compatible version; these are automatically detected and captured
+  elsewhere as described below.
+* `Environment markers`_ should be used as necessary to indicate dependencies
+  which should only be installed on specific operating systems, Python
+  versions, etc.
+* Avoid direct links to packages in local directories, GitHub, or other version
+  control systems if at all possible; all dependencies should be installed
+  from `PyPI`_.  If you think you're in one of the rare circumstances where
+  installing a package from a URL is appropriate, see the notes below on
+  `Installing Dependencies from URLs`_
+* If the dependencies in one context are a superset of those in another one,
+  do not repeat the dependencies.  Instead, explicitly include the file
+  produced by ``pip-compile`` for the smaller set of dependencies in the
+  requirements file for the larger set of dependencies. For example,
+  ``test.in`` often includes a line like the following to ensure that the same
+  versions of packages used in production for a service will also be used when
+  testing it:
+
+.. code-block:: python
+
+  -r base.txt               # Core dependencies of the service being tested
+
+If the repository contains a ``setup.py`` file defining a Python package, the
+base dependencies also need to be specified there.  These can be derived from
+``requirements/base.in`` with a simple Python function declared in
+``setup.py`` itself.  An example can be found in the
+`setup.py file for edx-completion`_.
+
+.. _requirements file: https://pip.readthedocs.io/en/1.1/requirements.html
+.. _cookiecutter-django-app: https://github.com/edx/cookiecutter-django-app/tree/master/%7B%7Bcookiecutter.repo_name%7D%7D/requirements
+.. _Environment markers: https://www.python.org/dev/peps/pep-0508/#environment-markers
+.. _PyPI: https://pypi.org/
+.. _setup.py file for edx-completion: https://github.com/edx/completion/blob/master/setup.py
+
+Generate Exact Dependency Specifications
+----------------------------------------
+
+Although we want to keep our manually edited requirements files very simple,
+we need a separate set of requirements files which list every single package
+needed for each usage context, with exact versions of each for reproducible
+test runs and consistent development and production environments.  We can
+generate these automatically using `pip-tools`_, which consists of two related
+utilities:
+
+* ``pip-compile`` generates a requirements file from one or more high-level
+  input requirements files, listing exact versions of every listed and
+  indirect dependency needed to satisfy the given constraints.
+* ``pip-sync`` ensures that the current virtualenv contains exactly (and only)
+  the packages listed in the given requirements files, installing, upgrading,
+  and uninstalling packages as needed.
+
+Open edX packages use an ``upgrade`` make target to use ``pip-compile`` to
```

Comment:
Is this, "do use," or "should use"?

Reply:
To be technical, "many already do use, and the rest probably should use".